wrapper/src/main/java/software/amazon/jdbc/dialect/DialectManager.java

```diff
@@ -0,0 +1,103 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.dialect;
+
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import software.amazon.jdbc.AwsWrapperProperty;
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.util.ConnectionUrlParser;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.RdsUtils;
+import software.amazon.jdbc.util.StringUtils;
+import software.amazon.jdbc.util.Utils;
+
+public class DialectManager implements DialectProvider {
+
+  protected static DialectProvider customDialectProvider;
+
+  public static final AwsWrapperProperty DIALECT = new AwsWrapperProperty(
+      "wrapperDialect", "",
+      "A unique identifier for the supported database dialect. ");
+
+  protected static final Map<String, Dialect> knownDialectsByCode =
+      new HashMap<String, Dialect>() {
+        {
+          put("mysql", new MysqlDialect());
+          put("pg", new PgDialect());
+          put("mariadb", new MariaDbDialect());
+          put("aurora-mysql", new AuroraMysqlDialect());
+          put("aurora-pg", new AuroraPgDialect());
+        }
+      };
+
+  private final RdsUtils rdsHelper = new RdsUtils();
+  private final ConnectionUrlParser connectionUrlParser = new ConnectionUrlParser();
+
+  public static void setCustomDialectProvider(final @NonNull DialectProvider dialectProvider) {
+    customDialectProvider = dialectProvider;
+  }
+
+  public static void resetCustomDialectProvider() {
+    customDialectProvider = null;
+  }
+
+  public Dialect getDialect(final String driverProtocol, final String url, final Properties props)
+      throws SQLException {
+
+    final String userDialectSetting = DIALECT.getString(props);
+    final Dialect userDialect = knownDialectsByCode.get(userDialectSetting);
+    if (userDialect != null) {
+      return userDialect;
+    }
+
+    if (StringUtils.isNullOrEmpty(driverProtocol)) {
+      return new UnknownDialect();
+    }
+
+    final List<HostSpec> hosts = this.connectionUrlParser.getHostsFromConnectionUrl(url, true);
+    if (Utils.isNullOrEmpty(hosts)) {
+      throw new SQLException(Messages.get("DialectManager.parsedListEmpty",
+          new Object[] {url}));
+    }
+    final String host = hosts.get(0).getHost();
+
+    if (driverProtocol.contains("mysql")) {
+      if (this.rdsHelper.isRdsDns(host)) {
+        return new AuroraMysqlDialect();
+      }
+      return new MysqlDialect();
+    }
+
+    if (driverProtocol.contains("postgresql")) {
+      if (this.rdsHelper.isRdsDns(host)) {
+        return new AuroraPgDialect();
```

Comment:
What if we want to support RDS ? Is there a way to detect Aurora ?

Reply:
If RDS is different from vanilla MySQL and it affects the logic of the driver then we'll need to have a separate dialect class. In such case isConnectionDialect() can be used to detect a particular dialect. isConnectionDialect() isn't in use but I expect we might need it some day.

Reply:
RDS will have the same DNS, but different failover logic. So we should probably find a different way.